spring-cloud-dataflow-admin/src/main/java/org/springframework/cloud/dataflow/admin/controller/RuntimeModulesController.java

```diff
@@ -0,0 +1,194 @@
+
+		public Assembler() {
+			super(RuntimeModulesController.class, ModuleStatusResource.class);
+		}
+
+		@Override
+		public ModuleStatusResource toResource(ModuleStatus entity) {
+			return createResourceWithId(entity.getModuleDeploymentId(), entity);
+		}
+
+		@Override
+		protected ModuleStatusResource instantiateResource(ModuleStatus entity) {
+			ModuleStatusResource resource = new ModuleStatusResource(entity.getModuleDeploymentId().toString(), entity.getState().name());
+			List<ModuleInstanceStatusResource> instanceStatusResources = new ArrayList<>();
+			InstanceAssembler instanceAssembler = new InstanceAssembler(entity);
+			List<ModuleInstanceStatus> instanceStatuses = new ArrayList<>(entity.getInstances().values());
+			Collections.sort(instanceStatuses, INSTANCE_SORTER);
+			for (ModuleInstanceStatus moduleInstanceStatus : instanceStatuses) {
+				instanceStatusResources.add(instanceAssembler.toResource(moduleInstanceStatus));
+			}
+			resource.setInstances(new Resources<>(instanceStatusResources));
+			return resource;
+		}
+	}
+
+	@RestController
+	@RequestMapping("/runtime/modules/{moduleId}/instances")
+	@ExposesResourceFor(ModuleInstanceStatusResource.class)
+	public static class InstanceController {
+
+		private final Collection<ModuleDeployer> moduleDeployers;
+
+		@Autowired
+		public InstanceController(Collection<ModuleDeployer> moduleDeployers) {
+			this.moduleDeployers = new HashSet<>(moduleDeployers);
+		}
+
+		@RequestMapping
+		public PagedResources<ModuleInstanceStatusResource> list(@PathVariable String moduleId,
+				PagedResourcesAssembler<ModuleInstanceStatus> assembler) {
+			ModuleDeploymentId moduleDeploymentId = ModuleDeploymentId.parse(moduleId);
+			for (ModuleDeployer moduleDeployer : moduleDeployers) {
+				ModuleStatus status = moduleDeployer.status(moduleDeploymentId);
+				if (status != null) {
+					List<ModuleInstanceStatus> moduleInstanceStatuses = new ArrayList<>(status.getInstances().values());
+					Collections.sort(moduleInstanceStatuses, INSTANCE_SORTER);
+					return assembler.toResource(new PageImpl<>(moduleInstanceStatuses), new InstanceAssembler(status));
+				}
+			}
+			throw new ResourceNotFoundException();
+		}
+
+		@RequestMapping("/{instanceId}")
+		public ModuleInstanceStatusResource display(@PathVariable String moduleId, @PathVariable String instanceId) {
+			ModuleDeploymentId moduleDeploymentId = ModuleDeploymentId.parse(moduleId);
+			for (ModuleDeployer moduleDeployer : moduleDeployers) {
+				ModuleStatus status = moduleDeployer.status(moduleDeploymentId);
+				if (status != null) {
+					ModuleInstanceStatus moduleInstanceStatus = status.getInstances().get(instanceId);
+					if (moduleInstanceStatus == null) {
+						throw new ResourceNotFoundException();
+					}
+					return new InstanceAssembler(status).toResource(moduleInstanceStatus);
+				}
+			}
+			throw new ResourceNotFoundException();
+		}
+
+	}
+
+	@ResponseStatus(HttpStatus.NOT_FOUND)
```

Comment:
It is an easy way to trigger a 404 from the controller.

The only other Exception handler we maintained from XD is generic 500 for generic Exception